03app_labomap/03app_labomap.c

```diff
@@ -0,0 +1,168 @@
+
+    //  3           2            1           0
+    // 1098 7654 3210 9876 5432 1098 7654 3210
+    // .... .... .... .... .... .... ...A AAAA A: PIN:    03 (P0.03)
+    // .... .... .... .... .... .... ..B. .... B: PORT:    0 (P0.03)
+    // C... .... .... .... .... .... .... .... C: CONNECT: 0=Connected
+    // 0xxx xxxx xxxx xxxx xxxx xxxx xx00 0011 
+    //    0    0    0    0    0    0    0    3 0x00000003
+    NRF_TWI0->PSEL.SCL            = 0x00000003;
+
+    //  3           2            1           0
+    // 1098 7654 3210 9876 5432 1098 7654 3210
+    // .... .... .... .... .... .... ...A AAAA A: PIN:    02 (P0.02)
+    // .... .... .... .... .... .... ..B. .... B: PORT:    0 (P0.02)
+    // C... .... .... .... .... .... .... .... C: CONNECT: 0=Connected
+    // 0xxx xxxx xxxx xxxx xxxx xxxx xx00 0010
+    //    0    0    0    0    0    0    0    2 0x00000002
+    NRF_TWI0->PSEL.SDA            = 0x00000002;
+
+    //  3           2            1           0
+    // 1098 7654 3210 9876 5432 1098 7654 3210
+    // AAAA AAAA AAAA AAAA AAAA AAAA AAAA AAAA A: FREQUENCY: 0x01980000==K100==100 kbps
+    NRF_TWI0->FREQUENCY           = 0x01980000;
+}
+
+void i2c_end(void) {
+    // Disable TWI
+    //  3           2            1           0
+    // 1098 7654 3210 9876 5432 1098 7654 3210
+    // .... .... .... .... .... .... .... AAAA A: ENABLE: 0=Disabled
+    // xxxx xxxx xxxx xxxx xxxx xxxx xxxx 0000 
+    //    0    0    0    0    0    0    0    0 0x00000000
+    NRF_TWI0->ENABLE              = 0x00000000;
+}
+
+void i2c_begin(void) {
+    // Enable TWI
+    //  3           2            1           0
+    // 1098 7654 3210 9876 5432 1098 7654 3210
+    // .... .... .... .... .... .... .... AAAA A: ENABLE: 5=Enabled
+    // xxxx xxxx xxxx xxxx xxxx xxxx xxxx 0101 
+    //    0    0    0    0    0    0    0    5 0x00000005
+    NRF_TWI0->ENABLE              = 0x00000005;
+}
+
+
+void i2c_send(uint8_t addr, uint8_t* buf, uint8_t buflen) {
+    NRF_TWI0->ADDRESS             = addr;
+
+    uint8_t i = 0;
+    NRF_TWI0->TXD                 = buf[i];
+    NRF_TWI0->EVENTS_TXDSENT      = 0;
+    NRF_TWI0->TASKS_STARTTX       = 1;
+    i++;
+
+    while(i<buflen) {
+        while(NRF_TWI0->EVENTS_TXDSENT==0);
+        NRF_TWI0->EVENTS_TXDSENT  = 0;
+        NRF_TWI0->TXD             = buf[i];
+        i++;
+    }
+
+    while(NRF_TWI0->EVENTS_TXDSENT==0);
+    NRF_TWI0->TASKS_STOP     = 1;
+}
+
+
+void i2c_read(uint8_t addr, uint8_t* buf, uint8_t buflen) {
+    NRF_TWI0->ADDRESS = addr;
+
+    NRF_TWI0->TASKS_STARTRX = 1;
+    for (uint8_t i = 0; i < buflen; i++) {
+        while (NRF_TWI0->EVENTS_RXDREADY == 0);
+        NRF_TWI0->EVENTS_RXDREADY = 0;
+        buf[i]                    = NRF_TWI0->RXD;
+    }
+    
+    NRF_TWI0->EVENTS_STOPPED = 0;
+    NRF_TWI0->TASKS_STOP = 1;
+    while (NRF_TWI0->EVENTS_STOPPED == 0);
+}
+
+void SHT31_readTempHumidity(float* temp, float* humidity) {
+    uint8_t data[2];
+    uint16_t temp_raw, humidity_raw;
```

Comment:
declare one variables per line, indent so that all variable names start on the same column